### ISSUE TITLE
Robust Accuracy of L2_DeepFool and L2_CW Do Not Change At ALL

### ISSUE THREAD

**Issue by WhiteLie98** [bethgelab/foolbox#669]
Hi, I was confused by the results of L2 DeepFool Attacks and L2 CW Attacks. I tested both, their accuracies are 89%, whatever I changed. The changing includes steps, overshoot, stepsize, and epsilons. The network I chose is VGG-19, and I trained it on the CIFAR-10 dataset. Hope someone can help me out, thanks a lot.

The main code I put in here:
```python
attack1 = fa.L2DeepFoolAttack(steps=50, candidates=5, overshoot=0.01, loss="logits")
attack2 = fa.L2CarliniWagnerAttack(binary_search_steps=5, steps=100, stepsize=0.01, abort_early=False)

_, adv_images, _ = attack1(fmodel, images, labels, epsilons=0.01)
class_acc = accuracy(fmodel, adv_images, labels) * 100
print(f"The Classification Accuracy is {class_acc:.2f} %", flush=True)

_, adv_images, _ = attack2(fmodel, images, labels, epsilons=0.01)
class_acc = accuracy(fmodel, adv_images, labels) * 100
print(f"The Classification Accuracy is {class_acc:.2f} %", flush=True)
```

The result always shows like:
![image](https://user-images.githubusercontent.com/39949809/158376482-0a03a8b8-583a-4a3e-90da-8ef7554600dc.png)

**Comment by WhiteLie98** [comment]
Ok, here is the code.

```
from foolbox import PyTorchModel, accuracy
import foolbox.attacks as fa

import os
import torch
from vgg import *  # https://github.com/chengyangfu/pytorch-vgg-cifar10
import utils  # Set Seed
import torchvision
from torchvision.transforms import ToTensor, Compose, Normalize

def main() -> None:
    target_path = "./model/model_best.pth.tar"
    data_path = "./data"
    seed=42
    utils.setup_seed(seed)
    
    # Get Target Model
    model = vgg19().to("cuda")
    if os.path.isfile(target_path):
        checkpoint_orig = torch.load(target_path, map_location=torch.device("cuda"))
        checkpoint = {"state_dict": {}}
        for k in checkpoint_orig["state_dict"].keys():
            checkpoint["state_dict"][k.replace("module.", "")] = checkpoint_orig[ 
                "state_dict"
            ][k]
        del checkpoint_orig  # This is to save memory space
        model.load_state_dict(checkpoint["state_dict"])
    model.eval()

    # Get Vaild Images
    val_dataset = torchvision.datasets.CIFAR10(
        root=data_path, 
        train=False, 
        download=False, 
        transform=Compose([ToTensor(), Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])])
    )
    val_dataloader = torch.utils.data.DataLoader(val_dataset, batch_size=100, shuffle=False, num_workers=4)
    val_iter = iter(val_dataloader)
    images, labels = next(val_iter)
    images, labels = images.to("cuda"), labels.to("cuda")

    epsilons = [
        0.00005, # 0
        0.0002,   # 1
        0.0005,   # 2
        0.002,     # 3
        0.005,     # 4
        0.01,       # 5
        0.03,       # 6
        0.05,       # 7
        0.1,         # 8
        0.2,         # 9
        0.3,         # 10
        0.5,         # 11
        1.0          # 12
    ]
    fmodel = PyTorchModel(model, bounds=(0, 1))

    attack1 = fa.L2DeepFoolAttack(steps=50, candidates=5, overshoot=epsilons[4], loss="logits")
    attack2 = fa.L2CarliniWagnerAttack(binary_search_steps=5, steps=100, stepsize=epsilons[4], abort_early=False)

   # FoolBox
    print("Start Generating FoolBox Attacks...", end = "", flush=True)
    raw_advs, adv_images, success = attack1(fmodel, images, labels, epsilons=epsilons[8])
    print("\r" + "The Generation of the FoolBox Attack is Done!", flush=True)

    class_acc = accuracy(fmodel, adv_images, labels) * 100
    print(f"The Classification Accuracy is {class_acc:.2f} %", flush=True)

    print("*"*50)

    print("Start Generating FoolBox Attacks...", end = "", flush=True)
    raw_advs, adv_images, success = attack2(fmodel, images, labels, epsilons=epsilons[8])
    print("\r" + "The Generation of the FoolBox Attack is Done!", flush=True)

    class_acc = accuracy(fmodel, adv_images, labels) * 100
    print(f"The Classification Accuracy is {class_acc:.2f} %", flush=True)

    return 0

if __name__ == "__main__":
    main()
```

**Comment by zimmerrol** [comment]
I looks like the data normalization is a part of your data loader. Thus, the samples passed to foolbox won't be in the `[0, 1]` value range and, thus, you might have to chose very different epsilon values. This might mean that epsilon is way too small and, hence, even the best parameters of the attack possible won't increase the attack success rate much.
To change this, remove the normalization from the data loading and set the `preprocessing` argument of the `foolbox.PyTorchModel`. Maybe this already solves your issue.

**Comment by WhiteLie98** [comment]
Thanks for providing the method, but the result (for DeepFool) still be 95% after I replaced the normalization with `preprocessing = dict(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], axis=-3)`. And after I deleted all the normalization, the result maintains as 36%. By the way, to test the DeepFool, I changed the values of steps, candidates, overshoot, and epsilons.


> I looks like the data normalization is a part of your data loader. Thus, the samples passed to foolbox won't be in the `[0, 1]` value range and, thus, you might have to chose very different epsilon values. This might mean that epsilon is way too small and, hence, even the best parameters of the attack possible won't increase the attack success rate much. To change this, remove the normalization from the data loading and set the `preprocessing` argument of the `foolbox.PyTorchModel`. Maybe this already solves your issue.

**Comment by WhiteLie98** [comment]
Thanks for providing the solution, I once thought that the `preprocessing` is doing the same thing as `Normalize()` function, so I didn't notice that is the problem. Now I tried `preprocessing = dict(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], axis=-3)`, and it worked. Thanks again!

> I looks like the data normalization is a part of your data loader. Thus, the samples passed to foolbox won't be in the `[0, 1]` value range and, thus, you might have to chose very different epsilon values. This might mean that epsilon is way too small and, hence, even the best parameters of the attack possible won't increase the attack success rate much. To change this, remove the normalization from the data loading and set the `preprocessing` argument of the `foolbox.PyTorchModel`. Maybe this already solves your issue.

**Comment by WhiteLie98** [comment]
By the way, I don't know whether is my program's problem or the attack is too strong, I found the visual effect of the adversarial example usually quite different from the original one.

The adversarial example is inversed by the following code, for the reason that the input has been preprocessed.
```python
def invertrans(imgs):
    invTrans = Compose([Normalize(mean = [0., 0., 0.], std = [1/0.229, 1/0.224, 1/0.225]), Normalize(mean = [-0.485, -0.456, -0.406], std = [1., 1., 1.])])
    return invTrans(imgs)
```

I test the Linf PGD attack which achieves the robust accuracy of 75%.
```python
    fa_attack = fa.LinfPGD()

    print("Start Generating BoundaryAttack Attacks...", end = "", flush=True)
    _, adv_images, _ = fa_attack(fmodel, images, labels, epsilons=3e-3)
    print("\r" + "The Generation of the BoundaryAttack Attack is Done!", flush=True)

    writer.add_image("Attack", invertrans(adv_images[0].cuda()), 1)
    writer.add_image("Original", images[0], 1)

    class_acc = accuracy(fmodel, adv_images.cuda(), labels) * 100
    print(f"The Classification Accuracy is {class_acc:.2f} %", flush=True)
```

And the PGD image shows like this (I change several times the parameters like steps and epsilons, but they all end up like this):
![image](https://user-images.githubusercontent.com/39949809/158591354-cff4e738-72da-48a7-a7aa-91413b2b1ffb.png)

While the original one shows like this:
![image](https://user-images.githubusercontent.com/39949809/158591678-ca973218-65dc-4920-a63a-b0e05cabbf71.png)

Have you ever met such a thing before? I'm just curious.

**Comment by zimmerrol** [comment]
Glad to hear that your original issue is solved. I cannot really follow why you use your `invertrans` function for visualizing the adversarial images found by the attack. The results of the attack should be in the same space as the images you passed to the attack, i.e. if the original images are in the `[0,1]` range, the adversarial images returned will also be in the `[0, 1]` domain in you don't have to transform them.

**Comment by WhiteLie98** [comment]
Thank you so much! I really appreciate it!
The reason why I reversed the adversarial images was that I used other CW attacks at first, and the generated pictures were very strange. However, the visual effect was much better after using reverse, so I thought it was the `Normalize()` function that caused the problem.
I was so stupid for believing that, but now I get it. Thanks a lot!

> Glad to hear that your original issue is solved. I cannot really follow why you use your `invertrans` function for visualizing the adversarial images found by the attack. The results of the attack should be in the same space as the images you passed to the attack, i.e. if the original images are in the `[0,1]` range, the adversarial images returned will also be in the `[0, 1]` domain in you don't have to transform them.

**Comment by WhiteLie98** [comment]
Hi, it's me again :blush:. 
I found a way to deal with the Normalize() from the [torchattacks](https://github.com/Harry24k/adversarial-attacks-pytorch/blob/master/demos/White%20Box%20Attack%20(ImageNet).ipynb) which add a normalization layer before the target model. 
I write this here in case someone using multiple attack libraries just like me but doesn't know how to deal with the normalization from the DataLoader. And if what I said is wrong, please correct it.


> Thanks for providing the solution, I once thought that the `preprocessing` is doing the same thing as `Normalize()` function, so I didn't notice that is the problem. Now I tried `preprocessing = dict(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], axis=-3)`, and it worked. Thanks again!
> 
> > I looks like the data normalization is a part of your data loader. Thus, the samples passed to foolbox won't be in the `[0, 1]` value range and, thus, you might have to chose very different epsilon values. This might mean that epsilon is way too small and, hence, even the best parameters of the attack possible won't increase the attack success rate much. To change this, remove the normalization from the data loading and set the `preprocessing` argument of the `foolbox.PyTorchModel`. Maybe this already solves your issue.

**Comment by WhiteLie98** [comment]
Hi, it's me again :blush:. 
I found a way to deal with the Normalize() from the [torchattacks](https://github.com/Harry24k/adversarial-attacks-pytorch/blob/master/demos/White%20Box%20Attack%20(ImageNet).ipynb) which add a normalization layer before the target model. 
I write this here in case someone using multiple attack libraries just like me but doesn't know how to deal with the normalization from the DataLoader. And if what I said is wrong, please correct it.


> Thanks for providing the solution, I once thought that the `preprocessing` is doing the same thing as `Normalize()` function, so I didn't notice that is the problem. Now I tried `preprocessing = dict(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], axis=-3)`, and it worked. Thanks again!
> 
> > I looks like the data normalization is a part of your data loader. Thus, the samples passed to foolbox won't be in the `[0, 1]` value range and, thus, you might have to chose very different epsilon values. This might mean that epsilon is way too small and, hence, even the best parameters of the attack possible won't increase the attack success rate much. To change this, remove the normalization from the data loading and set the `preprocessing` argument of the `foolbox.PyTorchModel`. Maybe this already solves your issue.
